core/utils/basic_tables_utils.py

```diff
@@ -20,7 +20,10 @@ def build_sa_using_dfs(node, sa: list[int]):
 
     build_sa_using_dfs(suffix_tree.root, suf_tab)
 
-    return suf_tab
+    # assuming '$' is greater than all other alphabet
```

Comment:
I think you will find that the unicode for $ is 36 while A is 64 (and C, G, and T are larger). So $ is *not* greater than a DNA symbol. You don't just fix this by moving the final sentinel string '$' from the first to the last index. If you have the string aa$ the suffix will be ordered $, a$, aa$ if $ is the smallest letter (so you get the normal lexicographical order of a < aa), but if $ is greater you get aa$, a$, $. Whether $ is smaller than all letters or greater than all letters affect a lot of the ordering. It doesn't matter for how we use suffix tree or array, except that shorter prefixes are moved after or before longer strings, but you do not change the value of $ by moving one element.

Also, `suf_tab[0:1]` is more naturally written as `suf_tab[0]`.